src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/ChangeManager.java

```diff
@@ -0,0 +1,77 @@
+package uk.gov.companieshouse.overseasentitiesapi.utils;
+
+import java.util.Objects;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+
+/**
+ * A utility class to compare and apply changes to a Psc object
+ *
+ * @param <T> The type that contains the data to be changed
+ */
+public class ChangeManager<T> {
+
+  private final T currentState;
+
+
+  /**
+   * @param currentState The object containing the data that has been changed
+   */
+  public ChangeManager(T currentState) {
+    this.currentState = currentState;
+  }
+
+  /**
+   * @param proposedData      The data from the user submission
+   * @param currentData       The data from the public / private api
+   * @param proposedConverter A function to convert the proposed data to the setter type. If they
+   *                          are the same datatype, use Function.identity()
+   * @param equalityPredicate A function to compare the proposed and current data. If they are the
+   *                          same datatype, use Objects::equals
+   * @param dataSetter        A function to set the data on the Psc object
+   * @param <P>               The type of the proposed data
+   * @param <C>               The type of the current data
+   * @param <S>               The type of the setter data
+   * @return true if the data has changed, false otherwise
+   */
+  public <P, C, S> boolean compareAndApply(
+      P proposedData,
+      C currentData,
+      Function<P, S> proposedConverter,
+      BiPredicate<P, C> equalityPredicate,
+      BiConsumer<T, S> dataSetter) {
+
+    if (proposedData != null && !equalityPredicate.test(proposedData, currentData)) {
+      dataSetter.accept(currentState, proposedConverter.apply(proposedData));
+      return true;
+    }
+
+    return false;
+  }
+
+
+  /**
+   * Compares proposed data to current data and applies the proposed data if they differ. If the
+   * proposed data is null or the same as the current data, no changes are made.
+   * <br>
+   * Note: This method assumes that the proposed and current data are the same type if not use the
```

Comment:
Err, this assumption may not be valid. If that can't be enforced by a type constraint, I'd suggest not using (removing) this method.

Reply:
Hey, this is enforced by a type constraint. Sorry, the comment was a little misleading.
If you try putting two different data types there it will cause a compiler error

![image](https://github.com/companieshouse/overseas-entities-api/assets/124157110/6d93c844-f8fd-4405-85b5-69eb2c72e1ec)